nominal/nominal.py

```diff
@@ -1,628 +1,316 @@
-            )
-        )
 
-        # Make POST request to upload data file to S3
-        resp = requests.post(
-            url=ENDPOINTS["file_upload"].format(get_base_url(), self.filename),
-            data=csv_file_buffer.read(),
-            params={"sizeBytes": csv_buffer_size_bytes},
-            headers=self.__get_headers(content_type="octet-stream"),
+def get_default_connection() -> NominalClient:
+    """Retrieve the default global connection to the Nominal platform.
+
+    Raises nominal.exceptions.NominalError if no global connection has been set.
+    """
+    global _default_connection
+    if _default_connection is None:
+        raise NominalError("No default connection set: initialize with `set_default_connection(base_url, token)`")
+    return _default_connection
+
+
+def upload_dataset_from_pandas(
+    df: pd.DataFrame,
+    name: str,
+    description: str,
+    timestamp_column: str,
+    timestamp_type: TimestampColumnType,
+    *,
+    wait_until_complete: bool = True,
+) -> Dataset:
+    """Create a dataset in the Nominal platform from a pandas.DataFrame."""
+    conn = get_default_connection()
+
+    # TODO(alkasm): use parquet instead of CSV as an intermediary
+
+    def write_and_close(df: pd.DataFrame, w: BinaryIO) -> None:
+        df.to_csv(w)
+        w.close()
+
+    with reader_writer() as (reader, writer):
+        # write the dataframe to CSV in another thread
+        t = Thread(target=write_and_close, args=(df, writer))
+        t.start()
+        dataset = conn.create_dataset_from_io(
+            reader,
+            name,
+            timestamp_column=timestamp_column,
+            timestamp_type=timestamp_type,
+            file_type=FileTypes.CSV,
+            description=description,
+        )
+        t.join()
+    if wait_until_complete:
+        dataset.poll_until_ingestion_completed()
+    return dataset
+
+
+def upload_dataset_from_polars(
+    df: pl.DataFrame,
+    name: str,
+    description: str,
+    timestamp_column: str,
+    timestamp_type: TimestampColumnType,
+    *,
+    wait_until_complete: bool = True,
+) -> Dataset:
+    """Create a dataset in the Nominal platform from a polars.DataFrame."""
+    conn = get_default_connection()
+
+    def write_and_close(df: pl.DataFrame, w: BinaryIO) -> None:
+        df.write_csv(w)
+        w.close()
+
+    with reader_writer() as (reader, writer):
+        # write the dataframe to CSV in another thread
+        t = Thread(target=write_and_close, args=(df, writer))
+        t.start()
+        dataset = conn.create_dataset_from_io(
+            reader,
+            name,
+            timestamp_column=timestamp_column,
+            timestamp_type=timestamp_type,
+            file_type=FileTypes.CSV,
+            description=description,
+        )
+        t.join()
+    if wait_until_complete:
+        dataset.poll_until_ingestion_completed()
+    return dataset
+
+
+def upload_dataset(
```

Comment:
if we do want a singular method for file upload (vs `upload_csv` / `upload_parquet`), it should have `file` in the name for consistency with the dataframe methods

Reply:
(ignore if you've updated this is updated already locally!)